tests/components/mqtt/test_discovery.py

```diff
@@ -256,37 +413,407 @@ async def test_discovery_integration_info(
     assert state.name == "Milk"
 
     assert (
-        "Component has already been discovered: binary_sensor bla, sending update from external application bla2mqtt, version: 1.1, support URL: https://bla2mqtt.example.com/support"
+        f"Component has already been discovered: binary_sensor {discovery_id}"
         in caplog.text
     )
 
 
 @pytest.mark.parametrize(
-    "config_message",
+    ("single_configs", "device_discovery_topic", "device_config"),
+    [(TEST_SINGLE_CONFIGS, TEST_DEVICE_DISCOVERY_TOPIC, TEST_DEVICE_CONFIG)],
+)
+async def test_discovery_migration_to_device_base(
+    hass: HomeAssistant,
+    device_registry: dr.DeviceRegistry,
+    mqtt_mock_entry: MqttMockHAClientGenerator,
+    tag_mock: AsyncMock,
+    single_configs: list[tuple[str, dict[str, Any]]],
+    device_discovery_topic: str,
+    device_config: dict[str, Any],
+) -> None:
+    """Test the migration of single discovery to device discovery."""
+    await mqtt_mock_entry()
+
+    # Discovery single config schema
+    for discovery_topic, config in single_configs:
+        payload = json.dumps(config)
+        async_fire_mqtt_message(
+            hass,
+            discovery_topic,
+            payload,
+        )
+    await hass.async_block_till_done()
+    await hass.async_block_till_done()
+
+    await help_check_discovered_items(hass, device_registry, tag_mock)
+
+    # Migrate to device based discovery
+    payload = json.dumps(device_config | {"migrate_discovery": True})
+    async_fire_mqtt_message(
+        hass,
+        device_discovery_topic,
+        payload,
+    )
+    await hass.async_block_till_done()
+    # Check we still have our mqtt items
+    await help_check_discovered_items(hass, device_registry, tag_mock)
```

Comment:
The integration that provides the entities and migrates them, is also resposible for cleaning up.